crates/sui-indexer-alt-consistent-store/src/db/mod.rs

```diff
@@ -0,0 +1,1170 @@
+    pub(crate) fn iter<J, K, V>(
+        &self,
+        checkpoint: u64,
+        cf: &impl AsColumnFamilyRef,
+        range: impl RangeBounds<J>,
+    ) -> Result<iter::FwdIter<'_, K, V>, Error>
+    where
+        J: Encode,
+    {
+        let IterBounds(lo, _, Some(mut inner)) = self.iter_raw(checkpoint, cf, range)? else {
+            return Ok(iter::FwdIter::new(None));
+        };
+
+        if let Some(lo) = &lo {
+            inner.seek(lo);
+        } else {
+            inner.seek_to_first();
+        }
+
+        Ok(iter::FwdIter::new(Some(inner)))
+    }
+
+    /// Create a reverse iterator over the values in column family `cf` at the given `checkpoint`,
+    /// optionally bounding the keys on either side by the given `range`. A reverse iterator yields
+    /// keys in descending bincoded lexicographic order.
+    ///
+    /// This operation can fail if the database does not have a snapshot at `checkpoint`.
+    pub(crate) fn iter_rev<J, K, V>(
+        &self,
+        checkpoint: u64,
+        cf: &impl AsColumnFamilyRef,
+        range: impl RangeBounds<J>,
+    ) -> Result<iter::RevIter<'_, K, V>, Error>
+    where
+        J: Encode,
+    {
+        let IterBounds(_, hi, Some(mut inner)) = self.iter_raw(checkpoint, cf, range)? else {
+            return Ok(iter::RevIter::new(None));
+        };
+
+        if let Some(hi) = &hi {
+            inner.seek_for_prev(hi);
+        } else {
+            inner.seek_to_last();
+        }
+
+        Ok(iter::RevIter::new(Some(inner)))
+    }
+
+    fn at_snapshot(&self, checkpoint: u64) -> Result<Arc<rocksdb::Snapshot<'_>>, Error> {
+        self.0.read().expect("poisoned").with(|f| {
+            let Some(snapshot) = f.snapshots.get(&checkpoint).cloned() else {
+                return Err(Error::NotInRange { checkpoint });
+            };
+
+            // SAFETY: Decouple the lifetime of the Snapshot from the lifetime of the
+            // RwLockReadGuard.
+            //
+            // The lifetime annotation on Snapshot couples its lifetime with the DB it came from,
+            // which is owned by `self` through `Inner`, so it is safe to extend the lifetime of
+            // the column family from that of the read guard, to that of `self` using `transmute`.
+            let snapshot: Arc<rocksdb::Snapshot<'_>> = unsafe { std::mem::transmute(snapshot) };
+
+            Ok(snapshot)
+        })
+    }
+
+    fn iter_raw<J>(
+        &self,
+        checkpoint: u64,
+        cf: &impl AsColumnFamilyRef,
+        range: impl RangeBounds<J>,
+    ) -> Result<IterBounds<'_>, Error>
+    where
+        J: Encode,
+    {
+        let s = self.at_snapshot(checkpoint)?;
+
+        let lo = match range.start_bound() {
+            Bound::Unbounded => None,
+            Bound::Included(start) => Some(key::encode(start)),
+            Bound::Excluded(start) => {
+                let mut start = key::encode(start);
+                if !key::next(&mut start) {
```

Comment:
Interesting. We need to do this here and on line 362 because rocksdb iter's lower bound is inclusive and upperbound is exclusive.